### ISSUE TITLE
SSD KV tier: successful hydrate does not skip prefill (prompt_cache_hits=0); enabling SSD loses the in-RAM prefix hit

### ISSUE THREAD

**Issue by Pushkinist** [Pushkinist/rMLX#9]
## What happened

With the SSD KV tier enabled (`--kv-ssd-cache-gb N`), a repeated long-prompt request **hydrates successfully from SSD but does not skip prefill** — the full prefill runs again and TTFT is unchanged.

From the server log (128k prompt, repeat request) the hydrate succeeds:

```
ssd-hydrate: RAM miss served from SSD tier  block_count=510 prefix_len=130560 bytes_read=2341553592
```

…but the same request then reports a cache miss and a full prefill:

```
generate: prompt-cache stats  prompt_cache_hits=0  prompt_cache_misses=N  prompt_cache_hit_rate=0.0
decode_profile  prefill_ms=342260  prompt_len=130810
```

So the hydrated KV (`prefix_len=130560`) is loaded but never registered as a prefix-cache hit, and the model re-runs the full ~130k-token prefill.

**Regression angle:** with SSD **off**, the same repeated 128k request hits the in-RAM automatic-prefix-cache (TTFT collapses from ~5 min to ~60 ms). Turning the SSD tier **on** causes the prefix to be spilled/evicted from RAM and then re-hydrated as a *miss* — so enabling SSD actually removes the prefill-skip that APC already provides in RAM.

## Reproduction

```sh
rmlx serve --model mlx-community/Qwen3.6-35B-A3B-8bit \
  --kv-quant rotor3 --max-ctx 140000 --kv-ssd-cache-gb 8 --device gpu
# send the same ~128k-token prompt twice; the 2nd request hydrates from SSD but prefill is not skipped
```

## Environment

- Model / arch: `mlx-community/Qwen3.6-35B-A3B-8bit`, `Qwen3_5MoeForConditionalGeneration`
- KV quant: rotor3, SSD tier on (`--kv-ssd-cache-gb 8`)
- rMLX version: 0.1.0
- macOS 26.4.1, Apple M5 Max

## What to fix

Wire the SSD hydrate result into the APC hit path so a hydrated prefix counts as a `prompt_cache_hit` and short-circuits prefill (as an in-RAM prefix match does). Trace `attach_ssd_tier` → `set_ssd_source` → `PromptCache::push_with_hydrate` and the generate-loop prefix lookup; ensure a hydrated entry is admitted into the in-RAM prefix index *before* the prefill decision, and that enabling SSD never downgrades an otherwise-available in-RAM hit.

**Comment by Pushkinist** [comment]
## Root cause (investigated)

The `ssd-hydrate: RAM miss served from SSD tier` fires correctly and hydrates a 510-block (~130,560-token) entry into RAM. But the hydrated entry's `prompt_token_ids` is the **block-aligned prefix only** (`prompt_ids[..block_count * BLOCK_TOKENS]`), because the `.kvb` block file stores `seq_len` but **not** the full prompt token IDs. So for a 130,810-token request the hydrated entry covers 130,560 tokens, not 130,810.

**Qwen3.5-MoE** uses the **`ExactOnly`** cache-reuse policy: the only arm that avoids a full re-prefill requires `entry.prompt_token_ids() == prompt_ids` (exact equality). `130560 != 130810` → the match falls through to `CacheLookup::Miss`, the hydrated KV is pushed into RAM and immediately discarded, and the full ~342 s prefill runs again. Hence `prompt_cache_hits=0` despite a successful hydrate.

**Gemma4 is NOT affected.** It uses the `Partial` policy, whose `is_strict_prefix_of(prompt_ids)` check returns `true` for a 130,560-token hydrated entry against a 130,810-token prompt (it *is* a strict prefix). Its SWA snapshot/restore path then restores the hydrated KV and re-prefills only the ~250-token tail. So on Gemma4 the SSD prefill-reuse benefit already works; the gap is specific to the MoE `ExactOnly` path.

This is **feature-sized**, not a one-line wiring bug: `ExactOnly` was designed for RAM-resident full snapshots, while the SSD tier recovers block-aligned prefixes — the two don't compose without either an I/O-format change or a new generate-loop code path.

## Fix options

### Option A — store full token IDs in `.kvb` (I/O schema change)
- `block_io::write_caches`: add a `token_ids` tensor (`u32`, shape `[seq_len]`) to the safetensors payload at spill time (the full `prompt_token_ids` is available there).
- `block_io::read_caches` + `HydratedBlock`: read and carry `full_prompt_ids`.
- Arch `SsdHydrate` impl: when present, set the hydrated entry's `prompt_token_ids` to the full IDs.
- **Pro:** the `ExactOnly` path needs zero changes — SSD-hydrated entries become indistinguishable from RAM entries.
- **Con:** ~520 KB extra per 130k-token entry in `.kvb`; a `.kvb` schema bump (pre-release, so old blocks can be treated as a graceful miss + delete, which block_io already supports).
- **Effort:** medium. Cleaner and more correct.

### Option B — SSD-hydrate-aware `Partial` path for MoE (generate-loop change)
- After a successful hydrate + `find_best_prefix`, detect the hydrated SSD entry and take a new path: treat the hydrated `kv_caches` **and** GDN `lin_caches` as the running state at position 130,560, then re-prefill only `prompt_ids[130560..]` (~250 tokens) on top of that state; emit the last-chunk logits as the first-token prediction.
- Likely needs a new `CacheLookup::SsdPrefix { kv_caches, lin_caches, prefix_len }` variant.
- **Pro:** no schema change, no storage overhead; conceptually unifies with the Gemma4 `Partial` behavior.
- **Con:** more generate-loop surgery. The existing Gemma4 `Partial` path handles KV truncation only; resuming MoE decode from a mid-prompt state with **existing GDN recurrent `lin_caches`** is subtler to get right (the GDN state has no sequence axis — snapshot/restore semantics, not truncation).
- **Effort:** medium-high. Riskier around the GDN recurrent state.

## Recommendation
**Option A** — it's the cleaner, lower-risk fix: the token IDs are already in hand at spill time, the storage cost is acceptable for the long-prompt SSD use case (the whole point is to avoid a multi-minute prefill), and it leaves the `ExactOnly` state machine untouched. Option B avoids the disk overhead but pushes the complexity into the MoE generate loop's GDN handling.

Relabeled as `enhancement` — this is a missing SSD-tier feature for the MoE `ExactOnly` path, not a regression.
